pkg/database/vercode.go

```diff
@@ -89,29 +91,47 @@ type VerificationCode struct {
 	// IssuingUserID is the ID of the user in the database that created this
 	// verification code. This is only populated if the code was created via the
 	// UI.
-	IssuingUserID uint `gorm:"column:issuing_user_id; type:integer;"`
+	IssuingUserID    uint  `gorm:"-"`
+	IssuingUserIDPtr *uint `gorm:"column:issuing_user_id; type:integer;"`
 
 	// IssuingAppID is the ID of the app in the database that created this
 	// verification code. This is only populated if the code was created via the
 	// API.
-	IssuingAppID uint `gorm:"column:issuing_app_id; type:integer;"`
+	IssuingAppID    uint  `gorm:"-"`
+	IssuingAppIDPtr *uint `gorm:"column:issuing_app_id; type:integer;"`
 
 	// IssuingExternalID is an optional ID to an external system that created this
 	// verification code. This is only populated if the code was created via the
 	// API AND the API caller supplied it in the request. This ID has no meaning
 	// in this system. It can be up to 255 characters in length.
-	IssuingExternalID string `gorm:"column:issuing_external_id; type:varchar(255);"`
+	IssuingExternalID string `gorm:"column:issuing_external_id; type:text;"`
 }
 
 // BeforeSave is used by callbacks.
 func (v *VerificationCode) BeforeSave(tx *gorm.DB) error {
+	v.IssuingUserIDPtr = uintPtr(v.IssuingUserID)
+	v.IssuingAppIDPtr = uintPtr(v.IssuingAppID)
+
+	if v.RealmID == 0 && v.Realm == nil {
+		v.AddError("realm_id", "is required")
+	}
+
 	if len(v.IssuingExternalID) > 255 {
 		v.AddError("issuingExternalID", "cannot exceed 255 characters")
 	}
 
-	if msgs := v.ErrorMessages(); len(msgs) > 0 {
-		return fmt.Errorf("validation failed: %s", strings.Join(msgs, ", "))
+	return v.ErrorOrNil()
+}
+
+// AfterFind runs after the verification code is found.
+func (v *VerificationCode) AfterFind(tx *gorm.DB) error {
+	v.IssuingUserID = uintValue(v.IssuingUserIDPtr)
+	v.IssuingAppID = uintValue(v.IssuingAppIDPtr)
+
+	if v.Realm == nil {
```

Comment:
do we really need to load the realm for every verification code read?

Reply:
Done